lib/Sema/ConstraintGraph.cpp

```diff
@@ -1558,6 +1558,123 @@ void ConstraintGraph::dump(llvm::raw_ostream &out) {
   print(CS.getTypeVariables(), out);
 }
 
+void ConstraintGraph::dumpActiveScopeChanges(llvm::raw_ostream &out,
+                                             unsigned indent) {
+  if (Changes.empty() && (ActiveScope->getStartIdx() == Changes.size()))
+    return;
+
+  // Collect Changes for printing.
+  std::map<TypeVariableType *, TypeBase *> tvWithboundTypes;
+  std::vector<TypeVariableType *> addedTypeVars;
+  std::vector<TypeVariableType *> equivTypeVars;
+  std::set<Constraint *> addedConstraints;
+  std::set<Constraint *> removedConstraints;
+  for (unsigned int i = ActiveScope->getStartIdx(); i < Changes.size(); i++) {
+    auto change = Changes[i];
+    switch (change.Kind) {
+    case ChangeKind::BoundTypeVariable:
+      tvWithboundTypes.insert(std::pair<TypeVariableType *, TypeBase *>(
+          change.Binding.TypeVar, change.Binding.FixedType));
+      break;
+    case ChangeKind::AddedTypeVariable:
+      addedTypeVars.push_back(change.TypeVar);
+      break;
+    case ChangeKind::ExtendedEquivalenceClass:
+      equivTypeVars.push_back(change.EquivClass.TypeVar);
+      break;
+    case ChangeKind::AddedConstraint:
+      addedConstraints.insert(change.TheConstraint);
+      break;
+    case ChangeKind::RemovedConstraint:
+      removedConstraints.insert(change.TheConstraint);
+      break;
+    }
+  }
+
+  // If there are any constraints that were both added and removed in this set
+  // of Changes, remove them from both.
+  std::set<Constraint *> intersects;
+  set_intersection(addedConstraints.begin(), addedConstraints.end(),
+                   removedConstraints.begin(), removedConstraints.end(),
+                   std::inserter(intersects, intersects.begin()));
+  llvm::set_subtract(addedConstraints, intersects);
+  llvm::set_subtract(removedConstraints, intersects);
+
+  // Print out Changes.
+  PrintOptions PO;
+  PO.PrintTypesForDebugging = true;
+  out.indent(indent);
+  out << "(Changes:\n";
+  if (!tvWithboundTypes.empty()) {
+    out.indent(indent + 2);
+    out << "(Newly Bound Type: \n";
```

Comment:
Nit: let's replace "Type" with "Variable(s)" or simplify it down to `Newly Bound:`